src/Microsoft.AspNet.Razor.Runtime/BufferEntryCollection.cs

```diff
@@ -0,0 +1,191 @@
+        {
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException("index");
+            }
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            if (value.Length - index < count)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+
+            while (count > 0)
+            {
+                // Split large char arrays into 1KB strings.
+                var currentCount = Math.Min(count, MaxCharToStringLength);
+                Add(new string(value, index, currentCount));
+                index += currentCount;
+                count -= currentCount;
+            }
+            _isModified = true;
+        }
+
+        /// <summary>
+        /// Adds an instance of <see cref="BufferEntryCollection"/> to the buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer collection to add.</param>
+        public void Add([NotNull] BufferEntryCollection buffer)
+        {
+            _buffer.Add(buffer.BufferEntries);
+            _isModified = true;
+        }
+
+        public void Clear()
+        {
+            _buffer.Clear();
+            _isModified = true;
+        }
+
+        /// <inheritdoc />
+        public IEnumerator<string> GetEnumerator()
+        {
+            return new BufferEntryEnumerator(_buffer);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        private sealed class BufferEntryEnumerator : IEnumerator<string>
+        {
+            private readonly Stack<IEnumerator<object>> _enumerators = new Stack<IEnumerator<object>>();
+            private readonly List<object> _initialBuffer;
+
+            public BufferEntryEnumerator(List<object> buffer)
+            {
+                _initialBuffer = buffer;
+                Reset();
+            }
+
+            public IEnumerator<object> CurrentEnumerator
+            {
+                get
+                {
+                    return _enumerators.Peek();
+                }
+            }
+
+            public string Current
+            {
+                get
+                {
+                    var currentEnumerator = CurrentEnumerator;
+                    Debug.Assert(currentEnumerator != null);
+
+                    return (string)currentEnumerator.Current;
+                }
+            }
+
+            object IEnumerator.Current
+            {
+                get
+                {
+                    return Current;
+                }
+            }
+
+            public bool MoveNext()
+            {
+                var currentEnumerator = CurrentEnumerator;
+                if (currentEnumerator.MoveNext())
+                {
+                    var current = currentEnumerator.Current;
+                    var buffer = current as List<object>;
```

Comment:
With how `.Current` is written, buffer should always be `null` unless I'm missing something.